File: android/src/main/kotlin/network/xyo/ble/xyo_ble/GattSingleRequest.kt
package network.xyo.ble.xyo_ble

import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import network.xyo.ble.flutter.protobuf.Gatt
import network.xyo.ble.generic.gatt.peripheral.XYBluetoothResult
import network.xyo.ble.generic.scanner.XYSmartScan

@kotlin.ExperimentalUnsignedTypes
class GattSingleRequest: GattRequestHandler() {

    companion object {
        fun process(smartScan: XYSmartScan, arguments: Any?, result: MethodChannel.Result): Deferred<Boolean> {
            return GlobalScope.async {
                val operation = operation(arguments) ?: return@async false
                val device = smartScan.devices[operation.deviceId] ?: return@async false

                val response: Gatt.GattResponse?
                var bleResult: XYBluetoothResult<Any>?= null
                device.connection {
                    bleResult = runCall(device, operation)
                    return@connection XYBluetoothResult(true)
                }
                response = response(operation, bleResult)
                relayResponse(response, result)
                cleanup(device, operation)
                return@async true
            }
        }
    }
}

File: ios/Classes/GattSingleRequest.swift
import Flutter
import XyBleSdk
import CoreBluetooth
import SwiftProtobuf

/// Makes a single gatt request
struct GattSingleRequest: GattRequestHandler {

    static func process(arguments: Any?, result: @escaping FlutterResult) {
        guard
            let operation = self.operation(arguments),
            let device = self.device(operation.deviceID)
            else {
                result(nil)
                return
        }

        var response: GattResponse?
        var bleResult: XYBluetoothResult?
        device.connection {
            bleResult = self.runCall(for: device, operation: operation)
        }.then {
            response = self.response(operation, response: bleResult)
            self.relayResponse(response, result: result)
        }.catch { error in
            response = self.response(operation, error: error)
            self.relayResponse(response, result: result)
        }.always {
            self.cleanup(for: device, operation: operation)
        }
    }

}
